src/kernel/fs.zig

```diff
@@ -0,0 +1,661 @@
+    var tree = testfs.tree.children.items[0];
+    testing.expect(tree.val.isFile());
+    testing.expectEqual(n, &tree.val.File);
+    testing.expect(std.mem.eql(u8, tree.name, "abc.txt"));
+    testing.expectEqual(tree.data, null);
+    testing.expectEqual(tree.children.items.len, 0);
+
+    // Creating a dir
+    var d = try openDir("/def", .CREATE_DIR);
+    testing.expectEqual(testfs.tree.children.items.len, 2);
+    tree = testfs.tree.children.items[1];
+    testing.expect(tree.val.isDir());
+    testing.expectEqual(d, &tree.val.Dir);
+    testing.expect(std.mem.eql(u8, tree.name, "def"));
+    testing.expectEqual(tree.data, null);
+    testing.expectEqual(tree.children.items.len, 0);
+
+    // Creating a file under a new dir
+    n = try openFile("/def/ghi.zig", .CREATE_FILE);
+    testing.expectEqual(testfs.tree.children.items[1].children.items.len, 1);
+    tree = testfs.tree.children.items[1].children.items[0];
+    testing.expect(tree.val.isFile());
+    testing.expectEqual(n, &tree.val.File);
+    testing.expect(std.mem.eql(u8, tree.name, "ghi.zig"));
+    testing.expectEqual(tree.data, null);
+    testing.expectEqual(tree.children.items.len, 0);
+
+    testing.expectError(Error.NoSuchFileOrDir, openDir("/jkl", .NO_CREATION));
+    testing.expectError(Error.NoSuchFileOrDir, openFile("/mno.txt", .NO_CREATION));
+    testing.expectError(Error.NoSuchFileOrDir, openFile("/def/pqr.txt", .NO_CREATION));
+    testing.expectError(Error.NoSuchFileOrDir, openDir("/mno/stu", .NO_CREATION));
+    testing.expectError(Error.NotADirectory, openFile("/abc.txt/vxy.md", .NO_CREATION));
+    testing.expectError(Error.IsADirectory, openFile("/def", .NO_CREATION));
+    testing.expectError(Error.InvalidFlags, openFile("/abc.txt", .CREATE_DIR));
+    testing.expectError(Error.InvalidFlags, openDir("/abc.txt", .CREATE_FILE));
+    testing.expectError(Error.NotAbsolutePath, open("", .NO_CREATION));
+    testing.expectError(Error.NotAbsolutePath, open("abc", .NO_CREATION));
+}
+
+test "read" {
+    var testfs = try testInitFs(std.heap.page_allocator);
+    root = testfs.tree.val;
+
+    var f = try openFile("/foo.txt", .CREATE_FILE);
+    var f_data = &testfs.tree.children.items[0].data;
+    var str = "test123";
+    f_data.* = try std.mem.dupe(std.heap.page_allocator, u8, str);
+    testing.expect(std.mem.eql(u8, str, try f.read(str.len)));
+    testing.expect(std.mem.eql(u8, str, try f.read(str.len + 1)));
+    testing.expect(std.mem.eql(u8, str, try f.read(str.len + 3)));
+    testing.expect(std.mem.eql(u8, str[0..0], try f.read(0)));
+}
+
+test "write" {
+    var testfs = try testInitFs(std.heap.page_allocator);
+    root = testfs.tree.val;
+
+    var f = try openFile("/foo.txt", .CREATE_FILE);
+    var f_data = &testfs.tree.children.items[0].data;
+    testing.expectEqual(f_data.*, null);
+
+    var str = "test123";
+    try f.write(str);
+    testing.expect(std.mem.eql(u8, str, f_data.* orelse unreachable));
+}
```

Comment:
Also need test for close.

Reply:
The thing is that closing can't really be tested because it's up to the underlying implementation to do stuff, like write to disk, sync over a network etc.

Reply:
I guess so, if it will only call the underlying impl and there won't be any high level code for closing, then shall be fine